client/rhel/rhn-client-tools/man/en/up2date.5
.\" Copyright 1999-2010 Red Hat, Inc.
.\"
.\" This man page is free documentation; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; version 2 of the License.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this man page; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH "UP2DATE" "5" "2010 August 30" "Linux" "Red Hat, Inc."
.SH NAME

up2date \- Configuration file for Spacewalk client programs.

.SH DESCRIPTION

.PP
Spacewalk client programs, such as \fBrhn_check(8)\fR, use a common
configuration file at \fI/etc/sysconfig/rhn/up2date\fR.

.SH OPTIONS

.IP \fBenableProxy\fR
Use a HTTP Proxy. Can be either 0 (False) or 1 (True).
.br
.IP \fBserverURL\fR
Remote server URL
.br
.IP \fBdebug\fR
Whether or not debugging is enabled. Can be either 0 (False) or 1 (True).
.br
.IP \fBsystemIdPath\fR
Location of system id file. Expects the path of a file on disk.
.br
.IP \fBversionOverride\fR
Override the automatically determined system version. Expects a string or blank (do not override the version).
.br
.IP \fBhttpProxy\fR
HTTP proxy in host:port format, e.g. squid.redhat.com:3128
.br
.IP \fBproxyUser\fR 
The username for an authenticated proxy.
.br
.IP \fBproxyPassword\fR
The password to use for an authenticated proxy.
.br
.IP \fBenableProxyAuth\fR
To use an authenticated proxy or not. Can be either 0 (False) or 1 (True).
.br
.IP \fBnetworkRetries\fR 
Number of attempts to make at network connections before giving up. Expects a
positive integer.
.br
.IP \fBsslCACert\fR
The CA cert used to verify the ssl server. Expects the path of a file on disk.
.br
.IP \fBnoReboot\fR
Disable the reboot action. Can be either 0 (False) or 1 (True).
.br
.IP \fBdisallowConfChanges\fR
Config options that can not be overwritten by a config update action. Expects a
semi-colon separated list of config option names.
.IP \fBhostedWhitelist\fR
Override the default list of URLs the RHN Register client treats as official RHN hosted servers. This only affects the RHN Register graphical- and text-based clients, as it changes some of the options that appear. This option is included for testing purposes only.
.br
.IP \fBretrieveOnly\fR
If set to 1 (True), package installs done through rhn_check will only download
the package, the package will not be installed.
.br
.SH FILES

.IP \fI/etc/sysconfig/rhn/up2date\fP
The location of this configuration file

.SH "SEE ALSO"

.PP
\fBrhn_check\fP(8), \fBrhn_register\fP(8), \fBrhnreg_ks\fP(8), \fBrhnsd\fP(8), \fBrhn-profile-sync\fP(8), \fBrhnplugin\fP(8), \fBrhnplugin.conf\fP(5).

.SH AUTHORS
.PP
See the AUTHORS file included with this software.
.PP
This manual page was written by James Bowes <jbowes@redhat.com>

.SH "BUGS"
.PP
Report bugs to <http://bugzilla.redhat.com>.

.SH COPYRIGHT

.PP
Copyright \(co 1999\-2010 Red Hat, Inc.

.PP
This is free software; see the source for copying conditions.  There is 
NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.


backend/server/handlers/xmlrpc/up2date.py
#
# Copyright (c) 2008--2012 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public License,
# version 2 (GPLv2). There is NO WARRANTY for this software, express or
# implied, including the implied warranties of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
# along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
# 
# Red Hat trademarks are not licensed under GPLv2. No permission is
# granted to use or replicate Red Hat trademarks that are incorporated
# in this software or its documentation. 
#
# Implements the up2date.* fucntions for XMLRPC
#

# system module import
import time
import string
import inspect

from rhn import rpclib
from types import ListType, TupleType, StringType, IntType
from spacewalk.common import rhnFlags, rhn_rpm
from spacewalk.common.rhnLog import log_debug, log_error
from spacewalk.common.rhnConfig import CFG
from spacewalk.common.rhnException import rhnFault
from spacewalk.common.rhnTranslate import _
from spacewalk.server.rhnLib import computeSignature
from spacewalk.server.rhnHandler import rhnHandler
from spacewalk.server import rhnChannel, rhnPackage, rhnDependency,\
    rhnCapability, rhnServer, rhnSQL
from spacewalk.server.rhnServer import server_route, server_hardware

import re
NONSUBSCRIBABLE_CHANNELS = re.compile("(rhn-proxy|rhn-satellite)")

class Up2date(rhnHandler):
    """ xml-rpc Server Functions that we will provide for the outside world.
    """
    def __init__(self):
        """ Up2date Class Constructor

           o Initializes inherited class.
           o Appends the functions available to the outside world in the
             rhnHandler list.
        """
        rhnHandler.__init__(self)
        # Function list inherited from rhnHandler
        # This action garners control of what is available to the client.

        # --- Clients v2+ ---
        # (getting headers, source and packages done with GETs now).
        self.functions.append('login')
        self.functions.append('listChannels')
        self.functions.append('reportSockets')
        self.functions.append('subscribeChannels')
        self.functions.append('unsubscribeChannels')
        self.functions.append('history')
        self.functions.append('solvedep')
        self.functions.append('solveDependencies')
        self.functions.append('solveDependencies_arch')
        self.functions.append('solveDependencies_with_limits')
    
    def auth_system(self, action, system_id):
        # Stuff the action in the headers:
        transport = rhnFlags.get('outputTransportOptions')
        transport['X-RHN-Action'] = action
        return rhnHandler.auth_system(self, system_id)

    def login(self, system_id, extra_data={}):
        """ Clients v2+
            Log in routine.
            Return a dictionary of session token/channel information.
            Also sets this information in the headers.
        """
        log_debug(5, system_id)
        # Authenticate the system certificate. We need the user record
        # to generate the tokens
        self.load_user = 1
        server = self.auth_system('login', system_id)
        # log the entry
        log_debug(1, self.server_id)
        # Update the capabilities list
        rhnCapability.update_client_capabilities(self.server_id)
        # Fetch the channels this client is subscribed to
        channelList = rhnChannel.channels_for_server(self.server_id)
        channels = []
        for each in channelList:
            if not each.has_key('last_modified'):
                # No last_modified attribute
                # Probably an empty channel, so ignore
                continue
            channel = [each['label'], each['last_modified']]
            # isBaseChannel
            if each['parent_channel']:
                flag = "0"
            else:
                flag = "1"
            channel.append(flag)

            # isLocalChannel
            if each['local_channel']:
                flag = "1"
            else:
                flag = "0"
            channel.append(flag)

            channels.append(channel)

        rhnServerTime = str(time.time())
        expireOffset = str(CFG.CLIENT_AUTH_TIMEOUT)
        signature = computeSignature(CFG.SECRET_KEY,
                                     self.server_id,
                                     self.user,
                                     rhnServerTime,
                                     expireOffset)

        loginDict = {
                'X-RHN-Server-Id'           : self.server_id,
                'X-RHN-Auth-User-Id'        : self.user,
                'X-RHN-Auth'                : signature,
                'X-RHN-Auth-Server-Time'    : rhnServerTime,
                'X-RHN-Auth-Expire-Offset'  : expireOffset,
                # List of lists [[label,last_mod],...]:
                'X-RHN-Auth-Channels'       : channels
                }

        # Duplicate these values in the headers so that the proxy can
        # intercept and cache them without parseing the xmlrpc.
        transport = rhnFlags.get('outputTransportOptions')
        for k, v in loginDict.items():
            # Special case for channels
            if string.lower(k) == string.lower('X-RHN-Auth-Channels'):
                # Concatenate the channel information column-separated
                transport[k] = map(lambda x: string.join(x, ':'), v)
            else:
                transport[k] = v
        log_debug(5, "loginDict", loginDict, transport)

        # store route in DB (schema for RHN 3.1+ only!)
        server_route.store_client_route(self.server_id)

        return loginDict


    def reportSockets(self, system_id, num_sockets):
        """ Clients v2+ """
        log_debug(5, system_id)
        # Authenticate the system certificate
        self.auth_system('reportSockets', system_id)
	server = rhnServer.search(self.server_id)
	server.reload_hardware()
	# there can only be one CPUDevice per system,
	# see hash fe46f64d
	[cpu_device] = server.hardware_by_class(server_hardware.CPUDevice)
	cpu_device.data["nrsocket"] = num_sockets
	cpu_device.status = 1 # mark as dirty
        # log the entry
        log_debug(1, "about to save %s" % self.server_id)
	cpu_device.save(self.server_id)
	rhnSQL.commit()
        log_debug(1, "saved %s" % self.server_id)
	return "reported %s sockets" % num_sockets


    def listChannels(self, system_id):
        """ Clients v2+ """
        log_debug(5, system_id)
        # Authenticate the system certificate
        self.auth_system('listChannels', system_id)
        # log the entry
        log_debug(1, self.server_id)
        channelList = rhnChannel.channels_for_server(self.server_id)
        return channelList


    def subscribeChannels(self, system_id, channelNames, username, passwd):
        """ Clients v2+ """
        log_debug(5, system_id, channelNames, username, passwd)
        # Authenticate the system certificate
        self.auth_system('subscribeChannel', system_id)
        # log the entry
        log_debug(1, self.server_id, channelNames)
        for channelName in channelNames:
            if NONSUBSCRIBABLE_CHANNELS.search(channelName):
                raise rhnFault(73, explain=False)
            else:
                rhnChannel.subscribe_channel(self.server_id, channelName,
                                         username, passwd)
        return 0


    def unsubscribeChannels(self, system_id, channelNames, username, passwd):
        """ Clients v2+ """
        log_debug(3)
        # Authenticate the system certificate
        self.auth_system('unsubscribeChannel', system_id)
        # log the entry
        log_debug(1, self.server_id, channelNames)
        for channelName in channelNames:
            rhnChannel.unsubscribe_channel(self.server_id, channelName,
                                           username, passwd)
        return 0

    def solvedep(self, system_id, deps):
        """ Clients v1-
            Solve dependencies for a given dependency problem list.
            IN:  a dependency problem list: [name, name, name, ...]
            RET: a package list: [[n,v,r,e],[n,v,r,e],...] That solves the
                 dependencies.
        """
        log_debug(4, system_id)
        return self.__solveDep(system_id, deps, action = "solvedep",
                               clientVersion = 1)


    def solveDependencies(self, system_id, deps):
        """ Clients v2+
            Solve dependencies for a given dependency problem list (newer version)
            IN:  a dependency problem list: [name, name, name, ...]
            RET: a hash {name: [[n, v, r, e], [n, v, r, e], ...], ...}
        """
        log_debug(4, system_id)
        return self.__solveDep(system_id, deps, action = "solvedep",
                               clientVersion = 2)

    def solveDependencies_arch(self, system_id, deps):
        """ Does the same thing as solve_dependencies, but also returns the architecture label with the
            package info.
            IN:  a dependency problem list: [name, name, name, ...]
            RET: a hash {name: [[n, v, r, e, a], [n, v, r, e, a], ...], ...}
        """
        log_debug(4, system_id)
        return self.__solveDep_arch(system_id, deps, action = "solvedep",
                               clientVersion = 2)

    def solveDependencies_with_limits(self, system_id, deps, all=0, limit_operator = None, limit = None):
        """ This version of solve_dependencies allows the caller to get all of the packages that solve a
            dependency and limit the packages that are returned to those that match the criteria defined
            by limit_operator and limit. This version of the function also returns the architecture label
            of the package[s] that get returned.

            limit_operator can be any of: '<', '<=', '==', '>=', or '>'.
            limit is a a string of the format [epoch:]name-version-release
            deps is a list of filenames that the packages that are returned must provide.
            version is the version of the client that is calling the function.
        """
        log_debug(4, system_id)
        return self.__solveDep_with_limits( system_id, deps, action = "solvedep", 
            clientVersion = 2, all=all, limit_operator=limit_operator, limit=limit)

    def history(self, system_id, summary, body = ""):
        """ Clients v2+
            Add a history log for a performed action
        """
        log_debug(5, system_id, summary, body)
        # Authenticate the system certificate
        server = self.auth_system('history', system_id)
        # log the entry
        log_debug(1, self.server_id)
        # XXX: Probably this should be a non fatal error...
        server.add_history(summary, body)
        server.save_history()
        return 0


    # --- PRIVATE METHODS ---

    def __solveDep_prepare(self, system_id, deps, action, clientVersion):
        """ Response for clients:
                version 1: list
                version 2: hash
        """
        log_debug(7, system_id, deps, action, clientVersion)
        faultString = _("Invalid value %s (%s)")
        if type(deps) not in (ListType, TupleType):
            log_error("Invalid argument type", type(deps))
            raise rhnFault(30, faultString % (deps, type(deps)))
        for dep in deps:
            if type(dep) is not StringType:
                log_error("Invalid dependency member", type(dep))
                raise rhnFault(30, faultString % (dep, type(dep)))
        # Ignore empty strings
        deps = filter(len, deps)
        # anything left to do?
        if not deps:
            return []
        # Authenticate the system certificate
        server = self.auth_system(action, system_id)
        log_debug(1, self.server_id, action, "items: %d" % len(deps))
        return deps

    def __solveDep(self, system_id, deps, action, clientVersion):
        """ Response for clients:
                version 1: list
                version 2: hash
        """
        log_debug(5, system_id, deps, action, clientVersion)
        result = self.__solveDep_prepare(system_id, deps, action, clientVersion)
        if result:
           # Solve dependencies
           result = rhnDependency.solve_dependencies(self.server_id,
                                                  result, clientVersion)
        return result

    def __solveDep_arch(self, system_id, deps, action, clientVersion):
        """ Response for clients:
                version 1: list
                version 2: hash
        """
        log_debug(5, system_id, deps, action, clientVersion)
        result = self.__solveDep_prepare(system_id, deps, action, clientVersion)
        if result:
            # Solve dependencies
            result = rhnDependency.solve_dependencies_arch(self.server_id,
                                                  result, clientVersion)
        return result


    def __solveDep_with_limits(self, system_id, deps, action, clientVersion, all=0, limit_operator=None, limit=None):
        """ Response for clients:
                version 1: list
                version 2: hash
        """
        log_debug(5, system_id, deps, action, clientVersion)
        result = self.__solveDep_prepare(system_id, deps, action, clientVersion)
        if result:
            # Solve dependencies
            result = rhnDependency.solve_dependencies_with_limits(self.server_id,
                                                  deps, clientVersion, all, limit_operator, limit)
        return result


class Servers(rhnHandler):
    """ A class to handle the site selection... """
    def __init__(self):
        """Servers Class Constructor. """
        rhnHandler.__init__(self)
        self.functions.append('get')
        self.functions.append('list')

    def get(self, *junk):
        """ Older funtion that can be a noop. """
        return []

    def list(self, systemid=None):
        """ Returns a list of available servers the client can connect to. """
        servers_list = [
            {
                'server'        :   'xmlrpc.rhn.redhat.com',
                'handler'       :   '/XMLRPC',
                'description'   :   'XML-RPC Server',
                'location'      :   'United States',
            },
        ]
        return servers_list

